add trade desk id support

lu-ruo opened the pull request.

lu-ruo commented on open-x-htb.js:
```diff
@@ -361,7 +362,7 @@ function OpenXHtb(configs) {
             partnerId: 'OpenXHtb',
             namespace: 'OpenXHtb',
-            statsId: 'OPNX',
+            statsId: 'OX',
```
According to targetingKeys definition: https://github.com/indexexchange/open-x-adapter/blob/master-v2/open-x-htb.js#L383 and the given unit tests: https://github.com/indexexchange/open-x-adapter/blob/master-v2/spec/profile.spec.js#L94 , the __profile.statsId should be 'OX' instead of 'OPNX'.
I also checked some calls to https://securepubads.g.doubleclick.net/gampad/ads? on some client's page and it is indicating ix_ox_om and ix_ox_id are used.
Please confirm this change is correct. And please confirm ix_ox_om and ix_ox_id are used as the targeting keys in client's lineitem setup so it can be triggered correctly.

lu-ruo commented on open-x-htb.js:
```diff
@@ -312,7 +312,8 @@ function OpenXHtb(configs) {
             if (bidDealId !== '') {
-                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pm] = [sizeKey + '_' + bidDealId];
+                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pm] = [sizeKey + '_' + bidPrice];
+                curReturnParcel.targeting[__baseClass._configs.targetingKeys.pmid] = [sizeKey + '_' + bidDealId];
```
According to your official documentation: https://github.com/lu-ruo/open-x-adapter/blame/master-v2/README.md#L235-L236, pm should be the bid cpm while pmid should be the bid deal id. The code here seems not correct. Please confirm this change wouldn't cause problems for current lineitem triggering.(aka, no client is using pm=<deal_id> as lineitem key-value targeting).

ix-certification replied:
The code in the documentation above is correct, however it is for a different template than OpenX is currently using (functionally, the templates are all the same). The changes you propose would cause the existing line items to stop triggering. We suggest that the changes be undone including the ones made to the `__profile.targetingKeys` object below.

lu-ruo replied:
Thanks for the clarification. I've restored the changes and added a comment in README.md.

lu-ruo commented on open-x-htb.js:
```diff
@@ -128,6 +129,31 @@ function OpenXHtb(configs) {
+              }
+            }
+          }
+          // identityData in beta format
```
Please advice if support for beta identity data format is needed.
FYI, in https://knowledgebase.indexexchange.com/display/PilotsPro/How+to+Use+adsrvr.org , search for
> (during beta, only the AdserverOrgIp.data.uid.id and AdserverOrgIp.data.uid.ext.rtiPartner objects are available).

lu-ruo commented on open-x-htb.js:
```diff
@@ -128,6 +129,31 @@ function OpenXHtb(configs) {
       ausString = ausString.slice(0, -1);
 
+      // process tdid query param
+      var identityData = returnParcels[0].identityData;
```
In your documentation https://knowledgebase.indexexchange.com/display/PilotsPro/How+to+Use+adsrvr.org , it's a bit vague where the identityData object would be -
> Within the returnParcels input/parameter, you can find all identity data available in the configuration under the identityData object.

Could you advice on which returnParcels 'input/parameter' it is talking about? - Currently, I'm assuming each returnParcel would have an identityData property. And if there are several returnParcel, I assume the identityData property would be the same for all of them.

lu-ruo commented on spec/support/libraryStubData.js:
```diff
@@ -44,7 +56,15 @@ var libraryStubData = {
     'openrtb.js': openRtbStub,
     'size.js': {
-        arrayToString: function (arr) {
+        arrayToString: function (arr, delim) {
```
There are 2 different signatures for this function in use. Updated the mock function for both cases.
https://github.com/indexexchange/open-x-adapter/blob/master-v2/open-x-htb.js#L125
https://github.com/indexexchange/open-x-adapter/blob/master-v2/open-x-htb.js#L315

lu-ruo commented on open-x-htb-system-tests.js:
```diff
@@ -0,0 +1,192 @@
+'use strict';
```
I've followed your guidelines here -https://knowledgebase.indexexchange.com/display/ADAPTER/Test+Case+Functions , but __generateRequestObj(returnParcels) in our adapter is never called within your test suite. I also used the example provided in the doc to check, but got failing test results for test case like 'should construct bid request correctly' because the request firing never got triggered. Could you provide some advice on this?

ix-certification replied:
The above issues could be related to the `getConfig` function in `open-x-htb-system-tests.js`. More information can be found [here](https://knowledgebase.indexexchange.com/display/ADAPTER/Test+Case+Functions#TestCaseFunctions-BidRequestTestCase).

1. The ad slot configurations need to be nested under the `xSlots` property (see example in link above).
2. You need to define only two items in `xSlots`, not three. 
3. The required partner-level property `host` is missing.
 
Once updated, it should look as follows:
```javascript
return {
    host: '<valid host>', // The required partner-level property.
    xSlots: { // All the ad slot-level properties under the `xSlots` property.
        1: {
            adUnitId: '54321',
            sizes: [[300, 250], [300, 600]]
        },
        2: {
            adUnitId: '12345',
            sizes: [[300, 600]]
        }
    }
};
```
 
This will allow the tests to run and for you to breakpoint in the adapter code. Failures that occur after the aforementioned changes are made will most likely be related to those tests cases that were written under the assumption that three items existed under `xSlots`. For example, in `validateBidRequest` it expects there to be three `auid`’s, however, there will only be two.

lu-ruo replied:
Thanks for the advice! After changing the `getConfig()` function, I'm able to get all the unit tests passing our expectation.

ix-certification commented:
A generic line item on our DFP instance works with all our partners. If you do not see an ad rendering it most likely has to do with the "Expected Bid / Deal ID" field under "Adapter Configuration > Ad Server". Information about the field can be found [here](https://knowledgebase.indexexchange.com/display/ADAPTER/Configuration#Configuration-AdServerSettings).
For your adapter, we were able to get the ads to render by mocking the OpenX endpoint locally to return `pub_rev: 1110` and setting "Expected Bid / Deal ID" to 110.

ix-certification commented on open-x-htb.js:
```diff
@@ -104,65 +105,79 @@ function OpenXHtb(configs) {
+        var gdprConsent = ComplianceService.gdpr.getConsent();
+        var gdprPrivacyEnabled = ComplianceService.isPrivacyEnabled();
+
+        var tradeDeskId;
```
Set `tradeDeskId` to `null` on declaration,  or this variable will retain the value from the previous iteration of the outer loop.

ix-certification commented on open-x-htb.js:
```diff
@@ -104,65 +105,79 @@ function OpenXHtb(configs) {
+
+        var identityData = returnParcels[0] && returnParcels[0].identityData;
+
+        if (identityData && identityData.AdserverOrgIp && identityData.AdserverOrgIp.data) {
```
If you are only expecting to have 1 `tradeDeskId` then you can `break` after setting `tradeDeskId`. If you are expecting more than one, then `tradeDeskId` should be instantiated as an empty array and then pushed into with the `id`s.

ix-certification commented on open-x-htb.js:
```diff
@@ -104,65 +105,79 @@ function OpenXHtb(configs) {
+            tws: Size.arrayToString([[Browser.getViewportWidth(), Browser.getViewportHeight()]]),
+            ifr: Browser.isTopFrame() ? 0 : 1,
+            callback: 'window.' + SpaceCamp.NAMESPACE + '.OpenXHtb.adResponseCallbacks.' + callbackId,
+            cache: new Date()
```
You can make use of the `System.now()` function outlined in the README.md. It implements the same functionality.

ix-certification commented on open-x-htb-system-tests.js:
```diff
@@ -0,0 +1,192 @@
+    expect(request.query.aus).toBe('300x250,300x600|300x600|728x90');
+}
+
+function getValidResponse(request, creative) {
```
Must ensure that for both `getValidBidResponse()` and `getValidBidResponseWithDeal()` that the winning bid’s `pub_rev` property is set such that it converts to 200 cents.

lu-ruo replied:
The thing I found out from unit test is - for `getValidBidResponse()` `pub_rev:1000` converts to `ix_ox_om:<size>_100`, but for `getValidBidResponseWithDeal()` `pub_rev:1000` converts to `ix_ox_pm:<size>_**1000**`, seems not using the same scale.
But as we are back to use the old template for openx you mentioned, ix_ox_pm would be: `<size>_<deal_id>`, so I guess the inconsistency in transformation is not related to us any more, just FYI.

ix-certification replied:
I am unable to reproduce the following behaviour.

>for getValidBidResponseWithDeal() pub_rev:1000 converts to ix_ox_pm:<size>_**1000**, seems not using the same scale.

The `pub_rev` value should not affect the `ix_ox_pm` key, it should only affect the `ix_ox_om` key.

As a test, I changed the `pub_rev` to `'1000'` and `'2000'` respectively in `getValidBidResponseWithDeal()` and then placed a breakpoint at the top of `validateTargetingWithDeal()`. Here is a screenshot of what I am seeing:

<img width="561" alt="screen shot 2018-10-18 at 4 27 35 pm" src="https://user-images.githubusercontent.com/28928533/47182195-dad34e80-d2f2-11e8-9cc5-14d819d33d8b.png">

Here the `ix_ox_om=300x250_100` is as I expected for a `pub_rev:'1000'`, and the `ix_ox_pm=300x250_1` for a `dealid:1`.

Could you please verify this item? If there is a difference in scaling between the Open Market and Private Market cases we will have to investigate further.

lu-ruo replied:
Hi @ix-certification I can verify what you describe is correct, using the current template for index, `pub_rev` only affects `ix_ox_pm `, so we are good.

What I was talking about before is - when I was trying to use the template described in README.md (the one you suggested me to undone, and I did undo it), I happened to find out the scale for `ix_ox_om` and `ix_ox_pm` is different,
![screen shot 2018-10-18 at 3 24 23 pm](https://user-images.githubusercontent.com/25375882/47187865-95ab1e80-d2ea-11e8-9762-259449d13463.png)
If you revert to commit 852285a and break at the same place. But I guess it's just how it is defined on your side, which is not related to the template OpenX is currently using.

ix-certification commented on open-x-htb-system-tests.js:
```diff
@@ -0,0 +1,192 @@
+    }));
+}
+
+function validatePixelRequests(pixelRequests) {
```
This test case is to test your adapter’s pixels, not ours. It also only supports pixels/beacons sent through XMLHttpRequests (XHR). The pixels sent by the OpenX adapter do not seem to use XHR, so this function can be removed.

lu-ruo commented:
Hi @ix-certification thanks a lot for your advice! Several things:
- After configuration changes (for both adapter-debugger page and in open-x-htb-system-tests.js), I'm able to verify openx adapter is working through the test page, and all the unit tests are passing our expectation now.
- I've addressed all the other comments related to code change.
- Regarding the private market targeting, I reverted to old ways of ix_ox_pm=<ad_unit_size>_<deal_id>,  kindly confirm this is the correct way for the existing line item targeting.

Let me know if you have other comments or this is ready for merge. Thanks!

ix-certification commented:
I've closed all but one discussion (the one to do with the CPM scaling issue you highlighted). Please have a look at that. Thanks.

lu-ruo commented:
https://github.com/indexexchange/open-x-adapter/pull/3#issuecomment-431156314 There might be a misunderstanding regarding that comment, I've explained a bit about it, hopefully it clears your concern. @ix-certification

ix-certification commented:
Ah, I see. Thanks for clearing that up. Everything else looks good. I'm going to merge in these changes.
